### PR TITLE
CT: Call op Berlin Revision

### PR DESCRIPTION
**Pull request by facuMH**
This PR expands on the existing call op to add support for the Berlin revision

This PR adds to #380

### CODE REVIEW

File: go/ct/spc/specification.go [hunk 1]
```diff
@@ -1787,32 +1787,45 @@ func rulesFor(i instruction) []Rule {
 	return res
 }
 
-func callOp() []Rule {
+func callRulesGenerator() []Rule {
```
Comment (HerbertJordan) [inline go/ct/spc/specification.go:1790]: `generateCallRules` or `getCallRules`?

File: go/ct/spc/specification.go [hunk 2]
```diff
@@ -1787,32 +1787,45 @@ func rulesFor(i instruction) []Rule {
...
+		for _, warm := range []bool{true, false} {
+			for _, zeroValue := range []bool{true, false} {
+				for _, static := range []bool{true, false} {
+					if rev == R07_Istanbul {
+						addressAccessCost = 0
+					} else if rev == R09_Berlin && warm {
+						addressAccessCost = 100
+					} else if rev == R09_Berlin && !warm {
+						addressAccessCost = 2600
+					}
```
Comment (HerbertJordan) [inline go/ct/spc/specification.go:1813]: To keep all gas-cost parameters together, I would move this part into the `callOp` instruction, similar to the `staticGas` selection code.

File: go/ct/spc/specification.go [hunk 3]
```diff
@@ -1900,13 +1929,17 @@ func callEffectIstanbul(s *st.State) {
...
 	// Deduct the gas costs for this call, except the costs for the recursive call.
-	dynamicGas, overflow := sumWithOverflow(memoryExpansionCost, positiveValueCost, valueToEmptyAccountCost)
+	dynamicGas, overflow := sumWithOverflow(memoryExpansionCost, positiveValueCost, valueToEmptyAccountCost, addrAccessCost)
 	if s.Gas < dynamicGas || overflow {
 		s.Status = st.Failed
 		return
 	}
 	s.Gas -= dynamicGas
 
+	if s.Revision == R09_Berlin {
```
Comment (HerbertJordan) [inline go/ct/spc/specification.go:1939]:
```suggestion
	if s.Revision >= R09_Berlin {
```

File: go/ct/utils/adapter.go [hunk 4]
```diff
@@ -72,8 +72,7 @@ type ctRunContext struct {
 // TODO: add unit test
 func (c *ctRunContext) AccountExists(addr vm.Address) bool {
 	existsAddr := c.state.Accounts.Exist(addr)
-	existsWarm := c.state.Accounts.IsWarm(addr)
-	return existsAddr || existsWarm
+	return existsAddr
```
Comment (HerbertJordan) [inline go/ct/utils/adapter.go:75]: nit: merge with previous line to `return c.state.Accounts.Exist(addr)`

**Review by HerbertJordan — APPROVED** [review]

**Review by simonlechner — DISMISSED** [review]
LGTM

File: go/ct/spc/specification.go [hunk 5]
```diff
@@ -1900,32 +1900,35 @@ func rulesFor(i instruction) []Rule {
...
+	revs := []Revision{R07_Istanbul, R09_Berlin, R10_London}
+	for _, rev := range revs {
+		for _, warm := range []bool{true, false} {
+			for _, zeroValue := range []bool{true, false} {
+				for _, static := range []bool{true, false} {
+					if static && !zeroValue {
+						res = append(res, callOp(rev, warm, zeroValue, callFailEffect, static)...)
+					} else {
+						res = append(res, callOp(rev, warm, zeroValue, callEffect, static)...)
+					}
```
Comment (HerbertJordan) [inline go/ct/spc/specification.go:1922]:
```suggestion
					effect := callEffect
					if static && !zeroValue {
						effect = callFailEffect
					}
					res = append(res, callOp(rev, warm, zeroValue, effect, static)...)
```

File: go/ct/spc/specification.go [hunk 6]
```diff
@@ -1900,32 +1900,35 @@ func rulesFor(i instruction) []Rule {
...
+				}
+			}
+		}
+	}
+
+	return res
+}
+
+func callOp(revision Revision, warm, zeroValue bool, opEffect func(s *st.State, addrAccessCost vm.Gas), static bool) []Rule {
```
Comment (HerbertJordan) [inline go/ct/spc/specification.go:1931]: Maybe rename this to `getRulesForCall`?

**Review by HerbertJordan — APPROVED** [review]

**Review by HerbertJordan — DISMISSED** [review]

**Review by simonlechner — APPROVED** [review]

**Pull request merged**